pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,355 @@
+
+	defer res.Close()
+
+	return &ftpFile{
+		response: res,
+		name:     s,
+		path:     filePath,
+		conn:     f.conn,
+		logger:   f.logger,
+		metrics:  f.metrics,
+	}, nil
+}
+
+// Mkdir creates a directory on the FTP server.
+// Here, os.FileMode is unused, but is added to comply with FileSystem interface.
+func (f *ftpFileSystem) Mkdir(name string, _ os.FileMode) error {
+	filePath := path.Join(f.config.RemoteDir, name)
+
+	defer f.processLog(&FileLog{Operation: "Mkdir", Location: filePath}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("Mkdir failed. Provide a valid directory : %v", errEmptyDirectory)
+		return errEmptyDirectory
+	}
+
+	err := f.conn.MakeDir(filePath)
+	if err != nil {
+		f.logger.Errorf("Mkdir failed. Error creating directory at %q : %v", filePath, err)
+		return err
+	}
+
+	f.logger.Logf("%s successfully created", name)
+
+	return nil
+}
+
+func (f *ftpFileSystem) mkdirAllHelper(filepath string) []string {
+	var dirs []string
+
+	currentdir := filepath
+
+	for {
+		err := f.conn.MakeDir(currentdir)
+		if err != nil {
+			parentDir, dir := path.Split(currentdir)
+
+			dirs = append([]string{dir}, dirs...)
+			if parentDir == "" || parentDir == "/" {
+				break
+			}
+			currentdir = path.Clean(parentDir)
+		} else {
+			dirs = append([]string{currentdir}, dirs...)
+			break
+		}
+
+	}
+
+	return dirs
+}
+
+// MkdirAll creates directories recursively on the FTP server. Here, os.FileMode is unused.
+// Here, os.FileMode is unused, but is added to comply with FileSystem interface.
+func (f *ftpFileSystem) MkdirAll(name string, _ os.FileMode) error {
+	defer f.processLog(&FileLog{Operation: "MkdirAll", Location: path.Join(f.config.RemoteDir, name)}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("MkdirAll failed. Provide a valid path : %v", errEmptyPath)
+		return errEmptyPath
+	}
+	// returns a slice of those directories that do not exist with the first index being the latest existing parent directory path.
+	dirs := f.mkdirAllHelper(name)
+
+	currentDir := dirs[0]
+
+	for i, dir := range dirs {
+		if i == 0 {
+			continue
+		}
+
+		currentDir = path.Join(currentDir, dir)
+
+		err := f.conn.MakeDir(currentDir)
+		if err != nil {
+			f.logger.Errorf("MkdirAll failed : %v", err)
+
+			return err
+		}
+	}
+
+	f.logger.Logf("Directories creation completed successfully.")
+
+	return nil
+}
+
+// Open retrieves a file from the FTP server and returns a file handle.
+func (f *ftpFileSystem) Open(name string) (container.File, error) {
+	filePath := path.Join(f.config.RemoteDir, name)
+
+	defer f.processLog(&FileLog{Operation: "Open", Location: filePath}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("Open_file failed. Provide a valid filename : %v", errEmptyFilename)
+		return nil, errEmptyFilename
+	}
+
+	res, err := f.conn.Retr(filePath)
+	if err != nil {
+		f.logger.Errorf("Open_file failed. Error opening file : %v", err)
+		return nil, err
+	}
+
+	_, s := path.Split(filePath)
```

Comment:
Now that this is simplified, this is just `s := path.Base(filePath)`.

I would pick a more descriptive name than `s` though. Something like `name` or at least `n` or `p`.

Actually, with `path.Base(filePath)` we can just throw this directly into the complex struct literal below.